neurodsp/aperiodic/dfa.py

```diff
@@ -0,0 +1,143 @@
+"""Fluctuation analyses to measure fractal properties of time series."""
+
+import numpy as np
+import scipy as sp
+
+from neurodsp.utils.data import split_signal
+
+###################################################################################################
+###################################################################################################
+
+def compute_fluctuations(sig, fs, n_scales=10, min_scale=0.01, max_scale=1.0, deg=1, method='dfa'):
+    """Compute a fluctuation analysis on a signal.
+
+    Parameters
+    ----------
+    sig : 1d array
+        Time series.
+    fs : float, Hz
+        Sampling rate, in Hz.
+    n_scales : int, optional, default=10
+        Number of scales to estimate fluctuations over.
+    min_scale : float, optional, default=0.01
+        Shortest scale to compute over, in seconds.
+    max_scale : float, optional, default=1.0
+        Longest scale to compute over, in seconds.
+    deg : int, optional, default=1
+        Polynomial degree for detrending. Only used for DFA.
+
+        - 1 for regular DFA
+        - 2 or higher for generalized DFA
+    method : {'dfa', 'rs'}
+        Method to use to compute fluctuations:
+
+        - 'dfa' : detrended fluctuation
+        - 'rs' : rescaled range
+
+    Returns
+    -------
+    t_scales : 1d array
+        Time-scales over which fluctuation measures were computed.
+    fluctuations : 1d array
+        Average fluctuation at each scale.
+    exp : float
+        Slope of line in loglog when plotting time scales against fluctuations.
+        This is the alpha value for DFA, or the Hurst exponent for rescaled range.
+
+    Notes
+    -----
+    These analyses compute fractal properties by analyzing fluctuations across windows.
+
+    Overall, these approaches involve dividing the time-series into non-overlapping
+    windows at log-spaced scales and computing a fluctuation measure across windows.
+    The relationship of this fluctuation measure across window sizes provides
+    information on the fractal properties of a signal.
+
+    Measures available are:
+
+    - DFA: detrended fluctuation analysis
+        - computes ordinary least squares fits across signal windows
+    - RS: rescaled range
+        - computes the range of signal windows, divided by the standard deviation
+    """
+
+    # Get log10 equi-spaced scales and translate that into window lengths
+    t_scales = np.logspace(np.log10(min_scale), np.log10(max_scale), n_scales)
+    win_lens = np.round(t_scales * fs).astype('int')
+
+    # Step through each scale and measure fluctuations
+    fluctuations = np.zeros_like(t_scales)
+    for idx, win_len in enumerate(win_lens):
+
+        if method == 'dfa':
+            fluctuations[idx] = compute_detrended_fluctuation(sig, win_len=win_len, deg=deg)
```

Comment:
This breaks when `win_len == 1`, causing `np.arange(win_len) ==  np.array([0]) ` to be the first arg in `np.polyfit`. Here is a minimal example:
```
import numpy as np
import scipy as sp
from neurodsp.utils.data import split_signal

tsig = np.random.randn(1000)

win_len = 1
segments = split_signal(np.cumsum(tsig - np.mean(tsig)), win_len).T
deg = 1

np.polyfit(np.arange(win_len), segments, deg=deg, full=True)
```

I'll have to dig into the code more to determine an appropriate fix for this.

Reply:
This seems like an edge case to me. DFA is just fitting a spline--that is, piecewise polynomials--to the data over intervals of length `win_len`. When `win_len ==1`, there's nothing to fit!

Reply:
Thanks, we can add this to fix:
```
if method == 'dfa' and win_len > 1:
```
I'll push this tonight.

Reply:
A window length of 1 is a bad settings, given the requested scales and sampling rate. I don't think we should be implicitly skipping this, but rather throwing an error for bad inputs. 

It's only an issue in our tests because we use a low sampling rate (100), which is lower than any real signals, so in practice this shouldn't often be an issue. 

I'm going to update the tests, since for these measures we can check real values, and that should avoid the testing problem.